src/btree/bt_cursor.c

```diff
@@ -585,6 +586,25 @@ __cursor_search_neighboring(WT_CURSOR_BTREE *cbt, WT_CURFILE_STATE *state, int *
     if (F_ISSET(cursor, WT_CURSTD_PREFIX_SEARCH))
         return (ret);
 
+    /*
+     * We walked off the end of the tree, we need to check records before our original position to
+     * see if they are visible. In snapshot isolation we can restart our search from where we first
+     * searched as no records will appear in between.
+     */
+    if (F_ISSET(session->txn, WT_TXN_HAS_SNAPSHOT)) {
+        __cursor_state_restore(cursor, state);
+        if (btree->type == BTREE_ROW)
+            WT_RET(__cursor_row_search(cbt, true, NULL, NULL));
```

Comment:
Given we've walked off the tree its possible our original page was evicted in the meantime and its structure has changed, as such our cbt->compare value may not be the same in this search as the previous search. In theory we should still see every record, i.e. cmp = 0, cmp > 0 and cmp < 0 as if compare could've been 0 it would've been in the original row/col search we did. This is somewhat hard to verify, I also imagine my explanation here isn't great.

Reply:
I don't know exactly what you mean by choosing cmp, as that is returned by the row_search. Or how we could set the cursor. My concern here is that we originally were placed >0 and then were placed < 0 or similar. Then in theory could we miss records in between? I can forward through the original compare value and check against it with an assertion? Another not so great option is preventing eviction of the original page while we do this search, then at least we skip row searching from the root page and can immediately start from where we left off.

Reply:
> My concern here is that we originally were placed >0 and then were placed < 0 or similar.

My suggestion was that we could deliberately make this happen with DIAGNOSTIC mode to provide confidence that it's not a problem.  So for example if we were originally placed > 0, then when we go back to the "original" position here, we deliberately move ourselves to < 0. I.e., if `cursor_row_search()` doesn't place us there, then we call `cursor->prev()` to get there. 

This is kind of similar to your suggestion that if we wind up < 0, we could move back to the original value and check with an assertion, except instead of validating correct behavior when a rare thing happens, it makes the rare thing more common so we can make sure it doesn't break things.

It's inelegant. So I'm not saying it's a good idea. Just offering it in the spirit of brainstorming.

Reply:
Oh right, yes that could work. Another possible albeit slightly higher complexity solution is traversing in the opposite direction to cmp originally. So if cmp is > 0 we walk `cursor->prev` first, thereby visiting the `cmp=0` area and vice versa. This may actually have make sense from a performance standpoint as we try to stay nearer to the original search location to begin with. That said it would be very workload dependent. @agorrod  what do you think about this issue and the proposed solutions?

Reply:
I have opted to go with my suggested solution of walking the opposite direction with respect to the side of the key we positioned on. That guarantees we see all the key range regardless of how the next search positions. However this change is getting a bit messy handling edge cases for prefix search near. So I will hold off on this change until prefix search near is deprecated in [WT-9142](https://jira.mongodb.org/browse/WT-9142).